### PR TITLE
Add step on refactoring for unit testability

### PR DESCRIPTION
**Pull request by bobcatfish**
The generator in `catgenerator.py` has several responsibilities
mixed together. It is unit testable but the resulting test
should be fairly hard to maintain. Breaking it up into a few
separate functions will make it easier to test and maintain.

@mtomwing

### CODE REVIEW

**Comment by bobcatfish** [comment]
> Coverage decreased

Sure did!

**Comment by bobcatfish** [comment]
Coverage has intentionally decreased, because I have added new untested code which students can add  unit tests for.

**Comment by bobcatfish** [comment]
Oh style checker, you are very particular about newlines.

File: catinabox/catgenerator.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        )
+
+
+def cat_generator():
+    while True:
+        try:
+            result = requests.get(NAME_GENERATOR_API_ENDPOINT)
+            name = result.json()[0]
+        except requests.exceptions.RequestException as e:
+            raise CouldNotGetNameError(e)
```
Comment (mtomwing) [inline catinabox/catgenerator.py:27]: What do you think about passing in `str(e)` instead? It doesn't look like `CouldNotGetNameError` needs anything in particular from the exception object except for that.

File: catinabox/catgenerator.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        current_time = int(time.time())
+        birthday = random.randint(
+            (current_time - (SECONDS_IN_YEAR * MAX_YEARS_OLD)),
```
Comment (mtomwing) [inline catinabox/catgenerator.py]: The outer pair of parentheses are unnecessary here.

File: catinabox/catgenerator.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            current_time)
+        birthday_datetime = time.strftime('%Y-%m-%d %H:%M:%S',
+                                          time.localtime(birthday))
+
+        yield {"name": name, "birthday": birthday_datetime}
```
Comment (mtomwing) [inline catinabox/catgenerator.py]: Maybe split the dict across multiple lines so it looks nicer.

**Comment by mtomwing** [comment]
Looks good except for my couple comments.

File: catinabox/catgenerator.py (hunk 1 above)
Comment (bobcatfish) [inline catinabox/catgenerator.py:27]: sounds good!

File: catinabox/catgenerator.py (hunk 2 above)
Comment (bobcatfish) [inline catinabox/catgenerator.py]: whoops!

File: catinabox/catgenerator.py (hunk 3 above)
Comment (bobcatfish) [inline catinabox/catgenerator.py]: lol okay

**Pull request merged**